nim.kak: General improvements

tawsi-melek opened the pull request:
- Add autoindentation for enum, tuple and object:

Before:
```nim
type
  MyType = tuple
  myint: int     # This line needs to be indented manually
```

After:
```nim
type
  MyType = tuple
    myint: int     # Lines after 'tuple' are now indented automatically
```

- Prevent keywords from being highlighted when inside of a comment (e.g. "if" was being highlighted when written in a comment)

- ~~Highlight Nim's escape sequences inside strings (\p,\n,\r,\t, etc.)~~ removed

- Add on/off to the boolean values understood by the highlighter

- Remove redundant regex highlighter for comments

- Fix autoindent indenting more than it should

mawww, as this is my first "proper" PR, I've included the Copyright Waiver.

SolitudeSF commented:
`enum`?

tawsi-melek commented:
Nice catch. Updated.

<details><summary>ot</summary>
<p>
It goes deeper than you think. I've been using Void Linux for years. Then I discovered nim two years ago (I'm amalek on the forum btw). And when some months ago I switched to kakoune from neovim I noticed you wrote the linter hook for nim and I said: "SolitudeSF? Wait a moment, I heard that name before..." Not that having Lain-chan as an avatar is that rare for programmers, mind you
</p></details>

SolitudeSF commented on rc/filetype/nim.kak:
```diff
@@ -37,8 +37,18 @@ add-highlighter shared/nim regions
+# Escape sequences in string
+add-highlighter shared/nim/string/escape regex \\[prcnlftv] 0:value
+# Comment tags
+add-highlighter shared/nim/comment/tags regex \b(TODO|FIXME|XXX|NOTE|BUG)\b 0:default+rb
```
This shouldn't be done by language plugin.

tawsi-melek replied:
Are you referring to both the escape sequences and the comment tags or just the tags? I know that there is [this](https://github.com/lenormf/kakoune-extra/blob/master/comnotes.kak) script for the tags, but it has a global scope and I couldn't get it to activate only on nim buffers (for some reason changing global/ to buffer/ completely disables ALL highlighters, even the ones in nim.kak).

Of course, if it's a kakoune policy I'll just remove it.

SolitudeSF replied:
these should be opt-in, since they are not related to language in any way. its a one liner, you can just add it to your own `hook global WinSetOption filetype=nim`.

tawsi-melek replied:
Ok. I removed them.

SolitudeSF commented on rc/filetype/nim.kak:
```diff
@@ -66,7 +76,7 @@ evaluate-commands %sh{
     float32 float64 bool char object seq array cstring string tuple varargs
     typedesc pointer byte set typed untyped void auto"
-    values="false true"
+    values="false true on off"
```
only place where on/off are used are pragmas for compiler options.

tawsi-melek replied:
No, they actually are aliases for true and false. I use them all the time. Try this on play.nim-lang.org ;-)
```nim
let is_on_an_alias_for_true = on

if is_on_an_alias_for_true != off:
  echo "YES!!!"
```

SolitudeSF replied:
thats cool that they are not even mentioned in manual or grammar. i guess that was a hack made specifically for pragmas.

tawsi-melek replied:
Yeah, Nim's manual is perennially lagging behind the actual specification.

tawsi-melek commented:
I've fixed a few additional things and added some more. Updated to OP.

SolitudeSF commented:
you've actually fixed some string/comment edgecases i struggled with. i have one left that i remember. `'#'` gets highlighted incorrectly.

tawsi-melek commented:
If you're referring to the fact that a single `#` wasn't enough to trigger the highlighter, that should be solved now.

I actually studied how python.kak and sh.kak did things and used their approach. One thing that would be cool to do is have formatted strings like `fmt"{myvar} is my var!"` highlight the sequence inside `{}`, but I'm still thinking about how to go about it.

SolitudeSF commented:
again, maybe overdoing is not the best. format strings are not builtin in nim, so you cant guarantee that `fmt"{myvar}"` in your code has the same meaning in someone else's code (even if its unlikely).

SolitudeSF commented:
also, this is how the `'#'` is displayed with your current version.
![image](https://user-images.githubusercontent.com/10168645/62734054-aae76a00-ba30-11e9-948d-e0f455197211.png)

tawsi-melek commented:
I've managed to fix that with the last commit, but there is still an edge case that I can't fix:

```nim
let
  c1 = '1' # This is highlighted properly
  c2 = '2'# This is NOT highlighted properly (notice the lack of space between ' and #)
```

SolitudeSF commented:
that probably pretty uncommon pattern. maybe you could take a look at another one:
![image](https://user-images.githubusercontent.com/10168645/62759476-14479700-ba8a-11e9-9ae0-9c92ab80888a.png)

this one is too, result of fixing another edgecase, but its not that uncommon.

SolitudeSF commented:
im not sure if this is even possible to fix.

i changed `nim/string` region highlighter to `add-highlighter shared/nim/string region '"' ((?<!\\)(\\\\)*"|$) group`, so `'"'` breaks only to the end of the line, but the edgecase i posted above is fixed then. fair tradeoff.
this goes out of my initial philosophy, that incorrect syntax shouldn't be highlighted, but i guess we're just trying to make it not break.

tawsi-melek commented:
I guess then we need a generalized char literal highlighter which overrides the string/comment hl. The c-family.kak seems to have one, I'll see if I can adapt it to work for Nim.

tawsi-melek commented:
Aaaand... it's done. I've modified the string hl after your input: it now works.

I've also ported the char hl from c-family.kak and made sure that the issue at https://github.com/mawww/kakoune/pull/3047#issuecomment-519683833 doesn't happen anymore.

If no more issues pop up, I think the PR can be merged.

SolitudeSF commented:
i guess a nitpick, but i made that only correct char literals are highlighted, now you can write incorrect char constanst and they will be highlighted. i guess more specification is more expensive, but was that really an issue?

tawsi-melek commented:
I reverted it. I changed it in the first place because I thought that it would fix https://github.com/mawww/kakoune/pull/3047#issuecomment-519683833, but upon close inspection all it was needed was to change the `comment_line`.
